src/Microsoft.SqlTools.ServiceLayer/DisasterRecovery/RestoreOperation/RestoreDatabaseService.cs

```diff
@@ -0,0 +1,217 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SqlServer.Management.Common;
+using Microsoft.SqlServer.Management.Smo;
+using Microsoft.SqlTools.ServiceLayer.Connection;
+using Microsoft.SqlTools.ServiceLayer.DisasterRecovery.Contracts;
+using Microsoft.SqlTools.ServiceLayer.TaskServices;
+
+namespace Microsoft.SqlTools.ServiceLayer.DisasterRecovery.RestoreOperation
+{
+    /// <summary>
+    /// Includes method to all restore operations
+    /// </summary>
+    public class RestoreDatabaseService
+    {
+
+        /// <summary>
+        /// Create a backup task for execution and cancellation
+        /// </summary>
+        /// <param name="sqlTask"></param>
+        /// <returns></returns>
+        internal async Task<TaskResult> RestoreTaskAsync(SqlTask sqlTask)
+        {
+            sqlTask.AddMessage(SR.Task_InProgress, SqlTaskStatus.InProgress, true);
+            RestoreDatabaseTaskDataObject restoreDataObject = sqlTask.TaskMetadata.Data as RestoreDatabaseTaskDataObject;
+            TaskResult taskResult = null;
+
+            if (restoreDataObject != null)
+            {
+                // Create a task to perform backup
+                return await Task.Factory.StartNew(() =>
+                {
+                    TaskResult result = new TaskResult();
+                    try
+                    {
+                        ExecuteRestore(restoreDataObject);
+                        result.TaskStatus = SqlTaskStatus.Succeeded;
+                    }
+                    catch (Exception ex)
+                    {
+                        result.TaskStatus = SqlTaskStatus.Failed;
+                        result.ErrorMessage = ex.Message;
+                        if (ex.InnerException != null)
+                        {
+                            result.ErrorMessage += System.Environment.NewLine + ex.InnerException.Message;
+                        }
+                    }
+                    return result;
+                });
+            }
+            else
+            {
+                taskResult = new TaskResult();
+                taskResult.TaskStatus = SqlTaskStatus.Failed;
+            }
+
+            return taskResult;
+        }
+
+       
+
+        /// <summary>
+        /// Async task to cancel restore
+        /// </summary>
+        public async Task<TaskResult> CancelTaskAsync(SqlTask sqlTask)
+        {
+            RestoreDatabaseTaskDataObject restoreDataObject = sqlTask.TaskMetadata.Data as RestoreDatabaseTaskDataObject;
+            TaskResult taskResult = null;
+
+
+            if (restoreDataObject != null)
```

Comment:
>restoreDataObject [](start = 16, length = 17)

do you also need to null check restoreDataObject.RestorePlan or is that property always non-null?